draft-ietf-quic-tls.md

```diff
@@ -667,73 +603,85 @@ Early data within the TLS connection MUST NOT be used.  As it is for other TLS
 
-### QHKDF-Expand
+## HelloRetryRequest
 
-QUIC uses the Hash-based Key Derivation Function (HKDF) {{!HKDF=RFC5869}} with
-the same hash function negotiated by TLS for key derivation.  For example, if
-TLS is using the TLS_AES_128_GCM_SHA256, the SHA-256 hash function is used.
+In TLS over TCP, the HelloRetryRequest feature ({{TLS13}; Section
+4.1.4) can be used to correct a client's incorrect KeyShare extension
+as well as for a stateless round trip check. From the perspective of
+QUIC, this just looks like additional messages carried in the Initial
+encryption level. Although it is in principle possible to use this
+feature for address verification in QUIC, QUIC implementations SHOULD
+instead use the Retry feature ({{QUIC-TRANSPORT}}; Section 4.4.2)).
 
-Most key derivations in this document use the QHKDF-Expand function, which uses
-the HKDF expand function and is modelled on the HKDF-Expand-Label function from
-TLS 1.3 (see Section 7.1 of {{!TLS13}}). QHKDF-Expand differs from
-HKDF-Expand-Label in that it uses a different base label and omits the Context
-argument.
 
-~~~
-QHKDF-Expand(Secret, Label, Length) =
-   HKDF-Expand(Secret, QhkdfExpandInfo, Length)
-~~~
+## TLS Errors
 
-The HKDF-Expand function used by QHKDF-Expand uses the PRF hash function
-negotiated by TLS, except for handshake secrets and keys derived from them (see
-{{handshake-secrets}}).
+If TLS experiences an error, it MUST generate an appropriate alert
+as defined in {{TLS13}}; Section 6) and then provide it to QUIC,
+which sends the alert in a CRYPTO_CLOSE frame. All such alerts are
+"fatal".
 
-Where the `info` parameter of HKDF-Expand is an encoded `QhkdfExpandInfo`
-structure:
 
-~~~
-struct {
-   uint16 length = Length;
-   opaque label<6..255> = "QUIC " + Label;
-} QhkdfExpandInfo;
-~~~
+# QUIC Packet Protection {#packet-protection}
+
+As with TLS over TCP, QUIC encrypts packets with keys derived from the TLS
+handshake, using the AEAD algorithm negotiated by TLS.
+
+
+## QUIC Packet Encryption Keys {#encryption-keys}
+
+QUIC derives packet encryption keys in the same way as TLS 1.3:
+Each encryption level/direction pair has a secret value, which
+is then used to derive the traffic keys using as described
+in {{TLS13}}; Section 7.3.
 
-For example, assuming a hash function with a 32 octet output, derivation for a
-client packet protection key would use HKDF-Expand with an `info` parameter of
-0x00200851554943206b6579.
+The keys for the Initial encryption level are computed based on
+the client's first Destination Connection Id, as described in
+{{initial-secrets}}.
 
+The keys for the remaining encryption level are computed in the same
+fashion as the corresponding TLS keys (see {{TLS13}}; Section 7),
+except that the label for HKDF-Expand-Label uses the prefix "quic "
+rather than "tls 13". The purpose of this change is to provide key
+separation between TLS and QUIC, so that TLS stacks can avoid
+exposing TLS record protection keys.
```

Comment:
I would like to see two clarifications here.

Am I correct in assuming that we would be using a modified version of HKDF-Expand-Label (that uses "quic " as the prefix) only when calculating the traffic keys? The other interpretation would be that we would be using the modified version for all cases (e.g. in Derive-Secret, when calculating the finished key). My preference goes to the former considering the fact that we are trying to change the traffic layer (i.e. TLS record) only.

The second issue is that the text seem to state that the non-modified version of HKDF-Expand-Label will be used for calculating the Initial Secret. I'd assume that it is a defect of the text.

Reply:
Glad that you raised this. I was assuming the former. I believe the latter would be safe but I would love to heard from @ad-l on this point.

The second is just a bug.

Reply:
Indeed, replacing all HKDF expansion labels in the TLS key schedule is the more conservative approach, as it prevents some secrets from being portable between TLS and QUIC. Most notably, if only the record key expansions use the "quic " label prefix, session tickets and PSK binders may potentially be re-used across the two protocols. /This is not necessarily a bad thing/ - a modular security model for the TLS handshake can in fact accommodate this a as feature and tolerate sessions that use a mix of record protocols. 

I would say it is up to the IETF to decide whether it wants the two protocols to be immiscible or not. If nobody sees value in mixed-transport sessions, going for disjoint labels is a cheap way to reduce the attack surface of both protocols and ensure they can be analyzed independently.

Reply:
@ekr @ad-l Thank you for the guidelines. The impact of sharing the labels are now clear to me.

To me now it seems that one way of making the decision in QUIC would be to follow the precedent that is soon going to be set by DTLS 1.3.

If DTLS allows session resumption using a ticket obtained through a TLS connection, I do not see why QUIC should disallow it. If DTLS uses the same traffic key as TLS does, I do not see why QUIC cannot use it as well.

Is my understanding correct that  DTLS 1.3 uses the same label for all the cases?

Reply:
DTLS 1.3 uses the same labels. 

@davidben the rationale here was less about analysis than about being to build a TLS 1.3 stack that guaranteed that it would keep the *TLS* keys internally. And this was intended as a cheap way to do that. I'm not religious about it, but there was some enthusiasm for it on the design team, so I suggested it.

Reply:
> the rationale here was less about analysis than about being to build a TLS 1.3 stack that guaranteed that it would keep the _TLS_ keys internally. And this was intended as a cheap way to do that. I'm not religious about it, but there was some enthusiasm for it on the design team, so I suggested it.

@ekr I remember having that discussion. OTOH, I do not recall discussing about _which_ (or that all) keys should be kept internally. It could be the case that the hesitation that some had was against sharing _one_ key between two stacks; an issue orthogonal to how we define the labels.

I think that we should discuss the issue in the working group to see what the concern is (or if such a concern really exists).

Reply:
@ekr Ah, that makes sense. I think we should try to avoid bifurcating all the labels but the "exported" record keys will need a dedicated path anyway, so it's hopefully not a huge deal either way.

Though having it lose the versioning is a little odd. Should it be a different input to hkdf-expand-label instead?